### DISCUSSION TITLE
Does OM-Compiler opimized for-loop?

### DISCUSSION THREAD

**Discussion by Arinomo** [OpenModelica/OpenModelica#11459] [Q&A] [answered]
Say I have two array `a[N]` and `b[N]` and want to calculate each element with for-loop:

```.mo
for idx in 1:N loop
  a[idx] = f(x, y, z);
end for;

for idx in 1:N loop
  b[idx] = f(a[idx], v, w);
end for;
```

Will the compiler optimized the code and put everything in one for-loop? or should I write everything in one for-loop from the beginning?

**Comment by perost** [comment, marked as answer]
For-equations are by default unrolled into individual equations, so in that case it doesn't matter. For-statements in an algorithm section will neither be unrolled nor merged though, though in the end we generate C code that the C compiler might optimize further.

For the new backend we're working on (compiler flag `--newBackend` to try it out) the focus is more on keeping arrays non-scalarized though, but even then it's not necessarily more optimal to put everything in one loop (the compiler might split them anyway).

**Reply by Arinomo** [reply]
thanks @perost!

**Comment by casella** [comment]
BTW, the RHS of this equation
```modelica
for idx in 1:N loop
  a[idx] = f(x, y, z);
end for;
```
is independent of `idx`. The backend has an optimization method `wrapFunctionCalls` that replaces each function call with an auxiliary variable, and then replaces all instances of that function call with that variable. Hence, that code fragment would be turned into something like the following pseudo-code:
```modelica
$cse1 = f(x,y,z);
for idx in 1:N loop
  a[idx] = $cse1;
end for;
```
This avoids multiple calls to the same function with the same arguments. Of course this optimization is only performed on [pure](https://specification.modelica.org/maint/3.6/functions.html#pure-modelica-functions) functions, which are guaranteed to always return the same result if given the same inputs.

**Reply by Arinomo** [reply]
@casella Thank you for the additional information about wrapping a function and about pure function. If I understand the documentation, any function that called an external "C" would (or should) be declared as `impure`, right? and it wont be optimized by the backend.

**Reply by perost** [reply]
External functions are assumed to be `impure` by default, though in newer versions of Modelica they are required to be explicitly declared `pure` or `impure`.

**Reply by casella** [reply]
Exactly. If you know that the C code always returns the same outputs for a given set of inputs, you can declare the function to be pure.

**Comment by casella** [comment]
In any case, even with the new array-preserving backend, I don't think there will be any difference between putting equations in two different loops with the same range, or in the same loop. The two models are completely equivalent, and they will produce the same compiled code.

**Reply by phannebohm** [reply]
That's correct. The new backend will always split for-loops, but "helping" the compiler by writing them separately is probably not going to make the compilation process any faster. Do what's best from a modeling/maintenance perspective!

**Comment by casella** [comment]
> Do what's best from a modeling/maintenance perspective!

That's the best advice when doing equation-based, object-oriented modelling 😃
